Nonpareil
=========

Estimate average coverage and create Nonpareil curves for metagenomic datasets.

[Stand-alone](http://nonpareil.readthedocs.org/) or
[Online querying](http://enve-omics.ce.gatech.edu/nonpareil/).

[![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat-square)](http://bioconda.github.io/recipes/nonpareil/README.html)
[![container ready](https://quay.io/repository/biocontainers/nonpareil/status)](https://quay.io/repository/biocontainers/nonpareil)

Citation
--------

If you use Nonpareil, please cite:

> Rodriguez-R *et al*. 2018. Nonpareil 3: Fast estimation of metagenomic
> coverage and sequence diversity. _mSystems_ 3(3): e00039-18.
> [DOI: 10.1128/mSystems.00039-18](https://doi.org/10.1128/mSystems.00039-18).
>
> Rodriguez-R & Konstantinidis. 2014. Nonpareil: a redundancy-based approach to
> assess the level of coverage in metagenomic datasets. _Bioinformatics_
> 30 (5): 629-635. [DOI: 10.1093/bioinformatics/btt584](https://doi.org/10.1093/bioinformatics/btt584).

For an extended discussion on coverage in metagenomic data, see also:

> Rodriguez-R & Konstantinidis. 2014. Estimating coverage in metagenomic data
> sets and why it matters. _The ISME Journal_ 8: 2349â€“2351.
> [DOI: 10.1038/ismej.2014.76](https://doi.org/10.1038/ismej.2014.76).
