<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Redundancy &#8212; Nonpareil 3.3.1rc2 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/classic.css?v=def86cc0" />
    
    <script src="_static/documentation_options.js?v=d1d09ded"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Nonpareil curves" href="curves.html" />
    <link rel="prev" title="Preprocess the reads" href="preprocess.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="curves.html" title="Nonpareil curves"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="preprocess.html" title="Preprocess the reads"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Nonpareil 3.3.1rc2 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Redundancy</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="redundancy">
<h1>Redundancy<a class="headerlink" href="#redundancy" title="Link to this heading">¶</a></h1>
<p>If you have large files (&gt;1Gb) and access to a cluster, take a look at <a class="reference internal" href="mpi.html"><span class="doc">MPI support</span></a>.</p>
<section id="for-the-impatient">
<h2>For the impatient<a class="headerlink" href="#for-the-impatient" title="Link to this heading">¶</a></h2>
<p>Even if you’re in a hurry, taking a look at <a class="reference internal" href="preprocess.html"><span class="doc">Preprocess the reads</span></a> is very important.
If you already did, you can simply run:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># fastq is recommended for kmer algorithm</span>
<span class="n">nonpareil</span> <span class="o">-</span><span class="n">s</span> <span class="n">reads</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">T</span> <span class="n">kmer</span> <span class="o">-</span><span class="n">f</span> <span class="n">fastq</span> <span class="o">-</span><span class="n">b</span> <span class="n">output</span>
<span class="n">nonpareil</span> <span class="o">-</span><span class="n">s</span> <span class="n">reads</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">T</span> <span class="n">kmer</span> <span class="o">-</span><span class="n">f</span> <span class="n">fasta</span> <span class="o">-</span><span class="n">b</span> <span class="n">output</span>

<span class="c1"># fasta is recommended for alignment algorithm</span>
<span class="n">nonpareil</span> <span class="o">-</span><span class="n">s</span> <span class="n">reads</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">T</span> <span class="n">alignment</span> <span class="o">-</span><span class="n">f</span> <span class="n">fasta</span> <span class="o">-</span><span class="n">b</span> <span class="n">output</span>
<span class="n">nonpareil</span> <span class="o">-</span><span class="n">s</span> <span class="n">reads</span><span class="o">.</span><span class="n">fa</span> <span class="o">-</span><span class="n">T</span> <span class="n">alignment</span> <span class="o">-</span><span class="n">f</span> <span class="n">fastq</span> <span class="o">-</span><span class="n">b</span> <span class="n">output</span>

<span class="c1"># gzipped files are also supported (with .gz extension)</span>
<span class="n">nonpareil</span> <span class="o">-</span><span class="n">s</span> <span class="n">reads</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span> <span class="o">-</span><span class="n">T</span> <span class="n">kmer</span> <span class="o">-</span><span class="n">f</span> <span class="n">fastq</span> <span class="o">-</span><span class="n">b</span> <span class="n">output</span>
</pre></div>
</div>
<p>Where <code class="docutils literal notranslate"><span class="pre">reads.fa</span></code> is the file containing the trimmed single reads, and
<code class="docutils literal notranslate"><span class="pre">output</span></code> is the prefix of the output files to be created.</p>
</section>
<section id="mandatory-options">
<h2>Mandatory options<a class="headerlink" href="#mandatory-options" title="Link to this heading">¶</a></h2>
<blockquote>
<div><dl class="option-list">
<dt><kbd><span class="option">-s <var>&lt;str&gt;</var></span></kbd></dt>
<dd><p>Path to the (input) file containing the sequences.
Gzipped files are supported with .gz extension</p>
</dd>
<dt><kbd><span class="option">-T <var>&lt;str&gt;</var></span></kbd></dt>
<dd><p>nonpareil algorithm. can be ‘kmer’ or ‘alignment’.</p>
</dd>
<dt><kbd><span class="option">-f <var>&lt;str&gt;</var></span></kbd></dt>
<dd><p>The format of the sequence: ‘fasta’ or ‘fastq’</p>
</dd>
</dl>
</div></blockquote>
</section>
<section id="common-options">
<h2>Common options<a class="headerlink" href="#common-options" title="Link to this heading">¶</a></h2>
<blockquote>
<div><dl class="option-list">
<dt><kbd><span class="option">-b <var>&lt;str&gt;</var></span></kbd></dt>
<dd><p>Path to the prefix for all the output files.  Replaces the options: -a, -C, -l, and -o; generating files
with the suffixes .npa, npc, .npl, and .npo, respectively, unless explicitly set.</p>
</dd>
<dt><kbd><span class="option">-d <var>&lt;num&gt;</var></span></kbd></dt>
<dd><p>Subsample iteratively applying this factor to the number of reads, resulting in logarithmic subsampling.
Use -d 0 to fall back to linear sampling, controlled by -m, -M, &amp; -i (this was the default before v2.4).
By default: 0.7.</p>
</dd>
<dt><kbd><span class="option">-n <var>&lt;int&gt;</var></span></kbd></dt>
<dd><p>Number of sub-samples to generate per point.  If it is not a multiple of the number of threads (see -t),
it is rounded to the next (upper) multiple.  By default: 1024.</p>
</dd>
<dt><kbd><span class="option">-L <var>&lt;num&gt;</var></span></kbd></dt>
<dd><p>Minimum overlapping percentage of the aligned region on the largest sequence. The similarity (see -S) is
evaluated for the aligned region only.  By default: 50.</p>
</dd>
<dt><kbd><span class="option">-X <var>&lt;int&gt;</var></span></kbd></dt>
<dd><p>Maximum number of reads to use as query.  This is capital X.  By default, 1,000 reads.</p>
</dd>
<dt><kbd><span class="option">-q <var>&lt;str&gt;</var></span></kbd></dt>
<dd><p>Path to the (input) file containing a second dataset to be used as query, for dataset comparisons.  This
option is currently experimental.</p>
</dd>
<dt><kbd><span class="option">-R <var>&lt;int&gt;</var></span></kbd></dt>
<dd><p>Maximum RAM usage in Mib.  Ideally this value should be larger than the sequences to analyze (discarding
non-sequence elements like headers or quality).  This is particularly important when running in multiple
cores (see -t).  This value is approximated.  By default 1024.
Maximum value in this version: 4194303</p>
</dd>
<dt><kbd><span class="option">-t <var>&lt;int&gt;</var></span></kbd></dt>
<dd><p>Number of threads.  Highest efficiency when the number of sub-samples (see -n) is multiple of the number
of threads.  By default: 2.</p>
</dd>
<dt><kbd><span class="option">-v <var>&lt;int&gt;</var></span></kbd></dt>
<dd><p>Verbosity level, for debugging purposes.  By default 7.  This is lowercase V.</p>
</dd>
<dt><kbd><span class="option">-V</span></kbd></dt>
<dd><p>Show version information and exit.  This is uppercase V.</p>
</dd>
<dt><kbd><span class="option">-h</span></kbd></dt>
<dd><p>Display this message and exit.</p>
</dd>
</dl>
</div></blockquote>
</section>
<section id="additional-options">
<h2>Additional options<a class="headerlink" href="#additional-options" title="Link to this heading">¶</a></h2>
<dl>
<dt><strong>Input/Output</strong></dt><dd><dl class="option-list">
<dt><kbd><span class="option">-a <var>&lt;str&gt;</var></span></kbd></dt>
<dd><p>Path to the (output) file where all data must be saved.  This report is not created by default.  See the
OUTPUT section.</p>
</dd>
<dt><kbd><span class="option">-C <var>&lt;str&gt;</var></span></kbd></dt>
<dd><p>Path to the (output) file where the mating vector is to be saved.  This is a capital C.</p>
</dd>
<dt><kbd><span class="option">-F</span></kbd></dt>
<dd><p>Report the sampled portions as a fraction of the library instead of the number of reads.  See -a, -o and
the OUTPUT section.</p>
</dd>
<dt><kbd><span class="option">-l <var>&lt;str&gt;</var></span></kbd></dt>
<dd><p>Path to the (output) file where the log of the run must be saved. By default the log is sent only to the
STDERR.  If set, the log is sent to both the STDERR and the log file.</p>
</dd>
<dt><kbd><span class="option">-o <var>&lt;str&gt;</var></span></kbd></dt>
<dd><p>Path to the (output) file where summary is to be saved.   By default the summary is sent to stdout (same
behavior as using a dash ‘-‘).  If an empty string ‘’ is provided, does not produce the summary. See the
OUTPUT section.</p>
</dd>
</dl>
</dd>
<dt><strong>Sampling</strong></dt><dd><dl class="option-list">
<dt><kbd><span class="option">-m <var>&lt;num&gt;</var></span></kbd></dt>
<dd><p>Minimum value of sampling portion.  By default: 0.</p>
</dd>
<dt><kbd><span class="option">-M <var>&lt;num&gt;</var></span></kbd></dt>
<dd><p>Maximum value of sampling portion.  By default: 1.</p>
</dd>
<dt><kbd><span class="option">-i <var>&lt;num&gt;</var></span></kbd></dt>
<dd><p>Interval between sampling portions. By default: 0.01.</p>
</dd>
</dl>
</dd>
<dt><strong>Mating</strong></dt><dd><dl class="option-list">
<dt><kbd><span class="option">-c</span></kbd></dt>
<dd><p>Do not use reverse-complement.  This is useful for single stranded sequences data (like RNA).  This is a
lowercase C.</p>
</dd>
<dt><kbd><span class="option">-N</span></kbd></dt>
<dd><p>Treat Ns as mismatches.  By default, Ns (unknown nucleotides) match any nucleotide (even another N).</p>
</dd>
<dt><kbd><span class="option">-S <var>&lt;num&gt;</var></span></kbd></dt>
<dd><p>Similarity threshold to group two reads together.   Reducing this option will increase sensitivity while
increasing running time.  This is uppercase S.</p>
</dd>
<dt><kbd><span class="option">-k <var>&lt;int&gt;</var></span></kbd></dt>
<dd><p>kmer size. You can increase kmer size to increase sensitivity. By default: 24</p>
</dd>
<dt><kbd><span class="option">-x <var>&lt;num&gt;</var></span></kbd></dt>
<dd><p>Probability of taking a sequence into account as query for the construction of the curve.  Higher values
reduce accuracy but increase speed.  This is lower case x.  If set, overides -X.</p>
</dd>
</dl>
</dd>
<dt><strong>Misc</strong></dt><dd><dl class="option-list">
<dt><kbd><span class="option">-A</span></kbd></dt>
<dd><p>Autoadjust parameters and re-run.  Evaluates the results looking for common problems, adjusts parameters
and re-run the analyses.  THIS IS EXPERIMENTAL CODE.</p>
</dd>
<dt><kbd><span class="option">-r <var>&lt;int&gt;</var></span></kbd></dt>
<dd><p>Random generator seed.  By default current time.</p>
</dd>
</dl>
</dd>
</dl>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Link to this heading">¶</a></h2>
<p>Sequences must be in FastA or FastQ format. See <a class="reference internal" href="preprocess.html"><span class="doc">Preprocess the reads</span></a>.</p>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Link to this heading">¶</a></h2>
<dl class="simple">
<dt>Redundancy summary: <code class="docutils literal notranslate"><span class="pre">.npo</span></code> file</dt><dd><p>Tab-delimited file with six columns. The first column indicates the sequencing effort (in number of reads), and the
remaining columns indicate the summary of the distribution of redundancy (from the replicates, 1,024 by default) at
the given sequencing effort. These five columns are: average redundancy, standard deviation, quartile 1, median
(quartile 2), and quartile 3.</p>
</dd>
<dt>Redundancy values: <code class="docutils literal notranslate"><span class="pre">.npa</span></code> file</dt><dd><p>Tab-delimited file with three columns. Similar to the .npo files, it contains information about the redundancy at
each sequencing effort, but it provides ALL the results from the replicates, not only the summary at each point. The
first column indicates the sequencing effort (as a fraction of the dataset), the second column indicates the ID of
the replicate (a number used only to introduce some controlled noise in plots), and the third column indicates the
estimated redundancy value.</p>
</dd>
<dt>Mates distribution: <code class="docutils literal notranslate"><span class="pre">.npc</span></code> file</dt><dd><p>Raw list with the number of reads in the dataset matching a query read. A set of query reads is randomly drawn by
Nonpareil (1,000 by default), and compared against all reads in the dataset. Each line on this file corresponds to a
query read (the order is not important). We have seen certain correspondance between these numbers and the distribution
of abundances in the community (compared, for example, as rank-abundance plots), but this file is provided only for
quality-control purposes and comparisons with other tools.</p>
</dd>
<dt>Log: <code class="docutils literal notranslate"><span class="pre">.npl</span></code> file</dt><dd><p>A verbose log of internal Nonpareil processing. The number to the left (inside squared brackets) indicate the CPU time
(in minutes). This file also provide quality assessment of the Nonpareil run (automated consistency evaluation). Ideally,
the last line should read “Everything seems correct”. Otherwise, it suggests alternative parameters that may improve the
estimation.</p>
</dd>
</dl>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <div>
    <h3><a href="index.html">Table of Contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">Redundancy</a><ul>
<li><a class="reference internal" href="#for-the-impatient">For the impatient</a></li>
<li><a class="reference internal" href="#mandatory-options">Mandatory options</a></li>
<li><a class="reference internal" href="#common-options">Common options</a></li>
<li><a class="reference internal" href="#additional-options">Additional options</a></li>
<li><a class="reference internal" href="#input">Input</a></li>
<li><a class="reference internal" href="#output">Output</a></li>
</ul>
</li>
</ul>

  </div>
  <div>
    <h4>Previous topic</h4>
    <p class="topless"><a href="preprocess.html"
                          title="previous chapter">Preprocess the reads</a></p>
  </div>
  <div>
    <h4>Next topic</h4>
    <p class="topless"><a href="curves.html"
                          title="next chapter">Nonpareil curves</a></p>
  </div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/redundancy.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="curves.html" title="Nonpareil curves"
             >next</a> |</li>
        <li class="right" >
          <a href="preprocess.html" title="Preprocess the reads"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Nonpareil 3.3.1rc2 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Redundancy</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    &#169; Copyright 2013-2017, Luis M Rodriguez-R.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.3.7.
    </div>
  </body>
</html>